### PR TITLE
API: Expose the dtype C API

### PR DESCRIPTION
**Pull request by ngoldbaum**
This exposes the DType API publicly. The functions are exposed using the codegen like the rest of the numpy API. The types were not straightofward to expose using the codegen, so I did it manually, keeping a separate PyCapsule containing a separate API table. Both the main array API table and the new dtype API table are exposed using `import_array()`.

### CODE REVIEW

**Comment by mattip** [comment]
Are all these APIs required to be made public in order to enable user-defined DTypes?

**Comment by mattip** [comment]
Could you divide this into two PRs:
- exposing APIs that can work with LIMITED_API
- exposing APIs that cannot work with LIMITED_API

**Comment by ngoldbaum** [comment]
Nice, it looks like that fix works, see last commit.

**Comment by ngoldbaum** [comment]
This should be ready to review as far as the code changes go. I still need to look at the documentation. I'm going to try to get the doxygen integration working (although that might be a pain because the codegen forces me to break the doxygen formatting slightly). If that doesn't work, I'll just add the docstrings to the sphinx API docs manually.

**Comment by ngoldbaum** [comment]
One question I have: how should I test that my changes to `import_array()` work when run against old versions of NumPy? Build an extension against NumPy 2 and run it against e.g. Numpy 1.26?

**Comment by ngoldbaum** [comment]
Excellent, tests are green :)

File: numpy/_core/include/numpy/experimental_dtype_api.h
Comment (ngoldbaum) [inline numpy/_core/include/numpy/experimental_dtype_api.h:321]: I never noticed until now but I think it's a bug that we exposed this name in the public DType API. I changed it to `PyArray_BytesDType` instead to avoid the python2 baggage, which is its internal name. This is being newly exposed so let's give it a name that makes sense in Python.

File: numpy/_core/include/numpy/dtype_api.h
Comment (ngoldbaum) [inline numpy/_core/include/numpy/dtype_api.h:58]: @seberg there are some comments below about "expected changes" for the ArrayMethod API. Do you still want to eventually do those changes? Now is probably the time...

File: numpy/_core/include/numpy/dtype_api.h
Comment (ngoldbaum) [inline numpy/_core/include/numpy/dtype_api.h:152]: `_NPY_METH_get_loop` is still marked as private. Maybe we should just make it public given it's pretty useful?

File: numpy/_core/include/numpy/dtype_api.h
Comment (ngoldbaum) [inline numpy/_core/include/numpy/dtype_api.h:199]: Should we give the other typedefs in this file names like `PyArrayMethod_ResolveDescriptorsFunction`?

File: numpy/_core/include/numpy/dtype_api.h
Comment (seberg) [reply numpy/_core/include/numpy/dtype_api.h:152]: yah, got to make it public, but we could follow up also to keep the diff smaller.

(There are a few things I dislike about it, but we aren't gonna change them quickly, and that is OK)

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:1643]: the distinction between these has always confused me so some copy I could stick in here would be really nice

**Comment by ngoldbaum** [comment]
The main thing left now is the typedefs, because I think we probably want to give them names in the `PyArray` namespace somewhere but wanted confirmation on that before doing it.

File: numpy/_core/src/multiarray/public_dtype_api.c
Comment (seberg) [inline numpy/_core/src/multiarray/public_dtype_api.c:177]: So I think I had suggested that before, but just mentionig it and hoping @mattip can voice an opinion:

An alternative that I rather like would be to add a big chunk of "empty slots" with a common to the api-gen (i.e. just with a `+ list(range(...))`) and then fill the normal API table here and above with that offset.

There isn't a huge reason for it, but I somehwat like the idea of having fewer structs.

---

I may have some tweaks if I review the API, but I itself, but I would prefer to just push this through, and the docs look nice on a quick glance.
(Tweaking this API is completely fine also after RC, the 0 people using it in their wheels already then, can recompile :))

File: doc/source/reference/c-api/array.rst
Comment (seberg) [inline doc/source/reference/c-api/array.rst:1643]: Feel free to shorten or rewrite, but a suggestion to get you started:
```suggestion
   One dimensional strided loops implementing the behavior (either a ufunc
   or cast).  In most cases, ``NPY_METH_strided_loop`` is the generic and only
   version that needs to be implemented.
   ``NPY_METH_contiguous_loop`` can be implemented additionally as a more
   light-weight/faster version and it is used when all inputs and outputs
   are contiguous.
   To deal with possibly unaligned data, NumPy needs to be able to copy
   unaligned to aligned data.  When implementing a new DType, the "cast" or
   copy for it needs to implement ``NPY_METH_unaligned_strided_loop ``.
   Unlike the normal versions, this loop must not assume that the data can be
   accessed in an aligned fashion.  These loops must copy each value before
   accessing or storing::
   
       type_in in_value;
       type_out out_value
       memcpy(&value, in_data, sizeof(type_in));
       out_value = in_value;
       memcpy(out_data, &out_value, sizeof(type_out)
   
   while a normal loop can just use `*(type_out *)out_data = *(type_in)in_data;`.
   The unaligned loops are currently only used in casts and will never be picked
   in ufuncs (ufuncs create a temporary copy to ensure aligned inputs).
   These loops are ignored when ``NPY_METH_get_loop`` is defined.
```
Plus, move `NPY_contiguous_index_loop` to later, and just say:  A specialized inner-loop to speed up common `ufunc.at` computations.  (But I would also be happy to leave it out.)

**Comment by seberg** [comment]
> One question I have: how should I test that my changes to import_array() work when run against old versions of NumPy? 

Don't worry about it too much, downstream runs these tests on their CI.

**Comment by ngoldbaum** [comment]
OK awesome. The last thing that needs to be documented is the typedefs. We have `PyArrayMethod_StridedLoop` already but all the other typedefs have names like `resolve_descriptors_function`. Do you agree that we should change the names of all the typedefs to use camel cased names in the `PyArrayMethod` namespace?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst]: p.s. I now see that @ngoldbaum already copied the suggestion - so this is really just to say that I agree!

**Comment by ngoldbaum** [comment]
I think this is fully ready now.

**Comment by ngoldbaum** [comment]
Also as a followup we're going to need to fix all the dtypes in numpy-user-dtypes after this is merged. We should probably also add a noisy cron CI job there to let us know if we break public API.

**Comment by ngoldbaum** [comment]
The doc build appears to have passed, but just a note that on my local mac setup I'm now hitting https://github.com/sphinx-doc/sphinx/issues/11449 when I do `spin docs` locally. Disabling parallel builds fixes it. There must be something in here that sphinx isn't happy about for whatever reason. The discussion in that issue is very confusing so I'm not sure what to do about it.

**Comment by ngoldbaum** [comment]
> just a note that on my local mac setup I'm now hitting https://github.com/sphinx-doc/sphinx/issues/11449 when I do spin docs locally

I ended up figuring out this was because of seg fault coming from IPython of all places, see https://github.com/sphinx-doc/sphinx/issues/11449#issuecomment-1942781601. Anyway, probably nothing to do with this PR.

**Review by seberg — APPROVED** [review]
I have pushed a small doc fixup and the second commit which removes the additional capsule and adds the new API directly (by just reserving slots 320-360 for these, the last of which is empty, but I thought it doesn't matter).

I also moved the table lookup defines into `_public_dtype_api_table.h` just to try to not bloat that `generate_api.py` script.

I think I prefer that quite a bit, not because it is much prettier code-wise, but because I don't like juggling multiple capsules.

---

I can see renaming some of the function definitions to include `Func` at the end (not the ones with `Loop`), and we should see if anyone has any other suggestions.

But for me: I think we can merge it now, ping the mailing list and create issues for anything else.  As this is completely new API, I do not see a problem with changing some things during RC phase, also.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1619]: These flags -> These functions

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1631]: This implies also the indexed loop is ignored, correct?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1646]: ufunc -> loop? (or does this not hold for casts?)

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1651]: ufunc -> loop

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1685]: what does "which is currently exposed publicly" mean here? I would have expected a (link to a) description of how to use this inside one's own resolution loop.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1702]: Insert "are" after "The *data*"

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1710]: Reorder to use fewer lines?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1776]: delete duplicate the

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1795]: Would be nice if, like for the other functions, this tells what the function should do! Probably can just adjust the text from below:
```
It is passed the operation and requested DType signatures and can mutate it to
attempt a new search for a matching loop/promoter.
```

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1829]: Confused what "Signal a floating point error respecting..." means. Does this *set* a floating point error? Or does it signal that such floating point errors should be caught?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1846]: Would be nice to give a concrete example...

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3256]: delete "is"

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3270]: This combination of flags and function typedef is rather nice. Use it for the array methods too?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3334]: I'm confused why this implements `PyArrayMethod_GetLoop` when above these were discussed under traverse loops.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3365]: Why would the general function be unsuitable?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3375]: Would it be possible to have a working link to the funtion dtypedef (here and below)?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3358]: Would be good to have a working link to `PyArray_ArrFuncs` here. Also, are these still necessary, or just there for backwards compatibility? (e.g., `dotfunc` seems pretty pointless, but `compare` likely useful)

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:3431]: iomatic -> idiomatic

File: doc/source/reference/c-api/types-and-structures.rst
Comment (mhvk) [inline doc/source/reference/c-api/types-and-structures.rst:51]: Why the `()`? (not new, I know)

File: doc/source/reference/c-api/types-and-structures.rst
Comment (mhvk) [inline doc/source/reference/c-api/types-and-structures.rst:853]: should *not* be relied on?

**Review by mhvk — COMMENTED** [review]
Had a look at the new documentation and mostly found some typos, but also some requests for clarification.

Overall, very nice!

I guess one todo item will be to ensure the same user dtypes are updated to the new names...

File: doc/source/reference/c-api/array.rst
Comment (seberg) [reply doc/source/reference/c-api/array.rst:1631]: Unfortunately not.  TBH, I wouldn't be unhappy to just not expose the indexed loop for now.

File: doc/source/reference/c-api/array.rst
Comment (seberg) [reply doc/source/reference/c-api/array.rst:3365]: Some places might call getitem in an arguably invalid way (passing `NULL` as the array).  I would hope NumPy got rid of all of them and things are fine, but I guess you might run into it somewhere.  Plus, optimize it.

(Not sure how much it makes sense to get into the weeds of it?)

File: doc/source/reference/c-api/array.rst
Comment (seberg) [reply doc/source/reference/c-api/array.rst:1846]: (Just to note, I would be fine to not expose this, but it is useful for units, as they should be able to use existing math.)
Not sure if it's easy to make something not very long.  We could point to the unit dtype in the numpy-dtypes repo but that isn't exactly documentation.

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:1619]: changed it to "slot IDs", which I think is a little more accurate than "functions".

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:1631]: Either way is fine, but the text should be correct!

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:3334]: thanks, this is a typo.

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:1846]: I'm going to link to an example usage in `metadatadtype`. I also noticed that I forgot to document `PyArrayMethod_TranslateGivenDescriptors` and `PyArrayMethod_TranslateLoopDescriptors` so will document those as well.

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:3375]: Unfortunately only the whole `PyArray_ArrFuncs` struct is documented but not the typedefs that make up the struct. Would you mind if I just link to the `ArrFuncs` doc at the top of this section and we can come back to documenting all the typedefs later?

File: doc/source/reference/c-api/types-and-structures.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/types-and-structures.rst:51]: no idea, deleted it

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:3358]: Added a note that this will likely change in the future.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:3375]: Sure, no problem!

How about adding a note here that these are still meant to be deprecated?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:3358]: ah, missed that - so ignore my request just now to add such a note!

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:1668]: reworded

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst]: reworded

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:1847]: ended up including a simpler example because linking to the metadatatype code is complicated given that it might change in the future and i'd rather not link to a git hash from today.

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:3372]: Rearranged the arraymethod docs.

File: doc/source/reference/c-api/array.rst
Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst]: I added some clarifying text incorporating sebastian's explanation.

**Comment by ngoldbaum** [comment]
The last commit responds to all of Marten's comments.

**Comment by seberg** [comment]
Should we get this in once things pass, just so I can rebase the other ones?  (Happy if there are follow up changes.)  Assuming you were happy with my API table reorganization.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:1847]: Super, that example really helps! (and shows my picture of what was happening was wrong...)

**Comment by ngoldbaum** [comment]
Yeah no worries about the reorganization, the only reason I didn't do it myself is because I had docs to write still

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1816]: Isn't this a `c:function::`?

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:3372]: Thanks, that helps the arraymethod part!

**Review by mhvk — COMMENTED** [review]
Beyond the question whether things are `c:type` or `c:function`, I think you answered/dealt with all my queries, thanks!

File: doc/source/reference/c-api/array.rst
Comment (seberg) [reply doc/source/reference/c-api/array.rst:1816]: I looked through a bit and it seemed right?  This is a function type definition (user defines something with that type), while `c:function::` seems correctly used for functions the user would call.

Comment (ngoldbaum) [reply doc/source/reference/c-api/array.rst:1816]: Yup, see https://www.sphinx-doc.org/en/master/usage/domains/c.html#directive-c-type. All the ones specified with `type` are function typedefs.

**Comment by ngoldbaum** [comment]
I missed a doc build warning which the last commit fixes, along with a reorganization that moves some old docs into a place that makes more sense, allowing me to add a new section heading that I can link to from another page.

File: doc/source/reference/c-api/array.rst
Comment (mhvk) [reply doc/source/reference/c-api/array.rst:1816]: Ah, sorry that i missed that!

**Comment by ngoldbaum** [comment]
@seberg did you want me to ping the mailing list before we merge this? or only after to alert everyone that it got merged?

**Comment by seberg** [comment]
I thought it's more useful aftr as the docs are build?

**Comment by ngoldbaum** [comment]
I could link to the doc build for this PR: https://output.circle-artifacts.com/output/job/39bc55ca-c9e3-42e5-940d-6224c4c903c8/artifacts/0/doc/build/html/reference/c-api/array.html#arraymethod-api

**Comment by mhvk** [comment]
My 2¢: merge and ask for comments then - even a major refactoring would at this stage be better done in a different PR!

**Comment by mhvk** [comment]
p.s. Probably good to note in any message to the mailing list that the new dtypes underlie the new `StringDType` as well as proper support for string ufuncs.

**Comment by ngoldbaum** [comment]
OK, fair enough, let's hit the merge button to unblock the other string PRs.

**Pull request merged**
